### PR TITLE
mtd: m25p80 spi driver update to support ACPI table match

### PR DESCRIPTION
**Pull request by msiddare**
Current m25p80 spi nor driver does not support ACPI table match.
Configuration done through ACPI tables is not recognized by driver.
Cisco 8000 platform cofigures NOR flash partition information
through APCI table which is not recognized by m25p80.

Added support in m25p80 to perform ACPI table match.

Signed-off-by: Madhava Reddy Siddareddygari <msiddare@cisco.com>

### CODE REVIEW

File: patch/cisco-acpi-spi-nor.patch [hunk 1]
```diff
@@ -0,0 +1,43 @@
+From a08c84878b4e26e04015737b5e85b0d90f8c1b51 Mon Sep 17 00:00:00 2001
+From: Madhava Reddy Siddareddygari <msiddare@cisco.com>
+Date: Tue, 8 Jun 2021 10:30:42 -0700
+Subject: [PATCH] mtd: m25p80 spi driver update to support ACPI table match
+
+Current m25p80 spi nor driver does not support ACPI table match.
+Configuration done through ACPI tables is not recognized by driver.
+Cisco 8000 platform cofigures NOR flash partition information
+through APCI table which is not recognized by m25p80.
```
Comment (paulmenzel) [inline patch/cisco-acpi-spi-nor.patch:9]: A*CP*I

File: patch/cisco-acpi-spi-nor.patch [hunk 2]
```diff
@@ -0,0 +1,43 @@
+From a08c84878b4e26e04015737b5e85b0d90f8c1b51 Mon Sep 17 00:00:00 2001
+From: Madhava Reddy Siddareddygari <msiddare@cisco.com>
+Date: Tue, 8 Jun 2021 10:30:42 -0700
+Subject: [PATCH] mtd: m25p80 spi driver update to support ACPI table match
+
+Current m25p80 spi nor driver does not support ACPI table match.
+Configuration done through ACPI tables is not recognized by driver.
+Cisco 8000 platform cofigures NOR flash partition information
```
Comment (paulmenzel) [inline patch/cisco-acpi-spi-nor.patch:8]: configures

File: patch/cisco-acpi-spi-nor.patch [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+through APCI table which is not recognized by m25p80.
+
+Added support in m25p80 to perform ACPI table match.
+
```
Comment (paulmenzel) [inline patch/cisco-acpi-spi-nor.patch:12]: Can you please provide the ASL code (fetch with `acpidump`, `acpixtract`, and `iasl -d`)? Please mention what firmware is used on the Cisco device.

Comment (msiddare) [reply patch/cisco-acpi-spi-nor.patch:12]: Please see ASL version

cisco@sonic:~$ iasl -d
Intel ACPI Component Architecture
ASL+ Optimizing Compiler/Disassembler version 20181213
Copyright (c) 2000 - 2018 Intel Corporation

ACPI dump
cisco@sonic:~$ sudo acpidump > acpi.dat
cisco@sonic:~$ ls
acpi.dat
cisco@sonic:~$ acpixtract -a acpi.dat 

Intel ACPI Component Architecture
ACPI Binary Table Extraction Utility version 20181213
Copyright (c) 2000 - 2018 Intel Corporation

  SSDT -  535404 bytes written (0x00082B6C) - ssdt1.dat
  MCFG -      60 bytes written (0x0000003C) - mcfg.dat
  EINJ -     336 bytes written (0x00000150) - einj.dat
  APIC -     230 bytes written (0x000000E6) - apic.dat
  SSDT -  933078 bytes written (0x000E3CD6) - ssdt2.dat
  TPM2 -      52 bytes written (0x00000034) - tpm2.dat
  BOOT -      40 bytes written (0x00000028) - boot.dat
  SLIT -     108 bytes written (0x0000006C) - slit.dat
  ERST -     560 bytes written (0x00000230) - erst.dat
  DSDT -   89633 bytes written (0x00015E21) - dsdt.dat
  SRAT -     216 bytes written (0x000000D8) - srat.dat
  UEFI -     566 bytes written (0x00000236) - uefi1.dat
  WDAT -     404 bytes written (0x00000194) - wdat.dat
  HEST -     168 bytes written (0x000000A8) - hest.dat
  BERT -      48 bytes written (0x00000030) - bert.dat
  SSDT -     758 bytes written (0x000002F6) - ssdt3.dat
  DMAR -     212 bytes written (0x000000D4) - dmar.dat
  FACP -     268 bytes written (0x0000010C) - facp.dat
  FPDT -      68 bytes written (0x00000044) - fpdt.dat
  SSDT -    1364 bytes written (0x00000554) - ssdt4.dat
  PRAD -     209 bytes written (0x000000D1) - prad.dat
  DBGP -      52 bytes written (0x00000034) - dbgp.dat
  HPET -      56 bytes written (0x00000038) - hpet.dat
  UEFI -      66 bytes written (0x00000042) - uefi2.dat
  FACS -      64 bytes written (0x00000040) - facs.dat
  BGRT -      56 bytes written (0x00000038) - bgrt.dat

cisco@sonic:~$ iasl -d ssdt1.dat 

Intel ACPI Component Architecture
ASL+ Optimizing Compiler/Disassembler version 20181213
Copyright (c) 2000 - 2018 Intel Corporation

Input file ssdt1.dat, Length 0x82B6C (535404) bytes
ACPI: SSDT 0x0000000000000000 082B6C (v02 CISCO  8800_RP  00000001 INTL 20181213)
Pass 1 parse of [SSDT]
Pass 2 parse of [SSDT]
Parsing Deferred Opcodes (Methods/Buffers/Packages/Regions)

Parsing completed
Disassembly completed
ASL Output:    ssdt1.dsl - 3191246 bytes

ASL code for the MTD partition

Please find attached ssdt1.dsl file which has ASL code for MTD partition
[ssdt1.zip](https://github.com/Azure/sonic-linux-kernel/files/6640629/ssdt1.zip)

Comment (paulmenzel) [reply patch/cisco-acpi-spi-nor.patch:12]: Thank you for your comment. Please edit your comment to mark up the paste as code block, indenting them with four spaces or tab or enclosing the block with ```, so it’s better readable.

**Comment by paulmenzel** [comment]
This depends on merge/pull request https://github.com/Azure/sonic-linux-kernel/pull/216/files does it? If so, please add this commit to merge/pull request https://github.com/Azure/sonic-linux-kernel/pull/216/files, so the dependencies are correct, and close this one.

**Comment by anshuv-mfst** [comment]
@lguohan - to help with merge, thanks

**Comment by msiddare** [comment]
This pull request does not depend on https://github.com/Azure/sonic-linux-kernel/pull/216/files, both of them can be merged independently.
sonic-linux-kernel/patch/series file update will take care of order of priority patches to be applied.
Both the patches can be applied independently and code does not depend on other pull request.

**Comment by paulmenzel** [comment]
Sorry for my error in reasoning regarding the dependencies. There is indeed no dependency.

The commits are still not squashed. Also, [according to the Linux kernel documentation](https://www.kernel.org/doc/html/latest/firmware-guide/acpi/enumeration.html#platform-bus-support), ACPI code should only be compiled in if `CONFIG_ACPI` is selected.

**Comment by msiddare** [comment]
CONIFG_ACPI is enabled by default in sonic-linux-kernel.

**Comment by paulmenzel** [comment]
> CONIFG_ACPI is enabled by default in sonic-linux-kernel.

That does not matter. The patches as is have to be in “upstream quality”. (And as written above *you* should preferably send them upstream too, so they get properly reviewed.)

**Comment by msiddare** [comment]
Hi paulmenzel,

This code doen't need to be under CONIFG_ACPI. As code is invoked only if acpi table are being passed to kernel.
.acpi_match_table = m25p_acpi_table

It is similar to .of_match_table = m25p_of_table, where code is not defined under CONFIG_OF option

**Comment by msiddare** [comment]
Review comments incorporated, ACPI code is compiled only if CONIFG_ACPI is enabled.

File: patch/cisco-acpi-spi-nor.patch [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
++};
++MODULE_DEVICE_TABLE(acpi, m25p_acpi_table);
++#endif
++
+ static struct spi_mem_driver m25p80_driver = {
+ 	.spidrv = {
+ 		.driver = {
+ 			.name	= "m25p80",
+ 			.of_match_table = m25p_of_table,
++			.acpi_match_table = m25p_acpi_table,
```
Comment (paulmenzel) [inline patch/cisco-acpi-spi-nor.patch:40]: The examples and code in the Linux kernel use `ACPI_PTR(m25p_acpi_table)`.

    $ git grep -C2 acpi_match_table

Why don’t you?

**Review by paulmenzel — COMMENTED** [review]
Thank you for improving the patch. Have you sent it upstream for review as talked about last week?

File: patch/cisco-acpi-spi-nor.patch (hunk 4 above)
Comment (msiddare) [reply patch/cisco-acpi-spi-nor.patch:40]: > Thank you for improving the patch. Have you sent it upstream for review as talked about last week?

Hi Paul,

I haven't sent this patch to upstream yet. First stage we are focusing on getting changes integrated to sonic-linux-kernel. Second stage will be to send patches upstream for review.

Comment (msiddare) [reply patch/cisco-acpi-spi-nor.patch]: > ACPI_PTR(m25p_acpi_table)

Addressed as per review comments.

File: patch/cisco-acpi-spi-nor.patch [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+--- a/drivers/mtd/devices/m25p80.c
++++ b/drivers/mtd/devices/m25p80.c
+@@ -319,11 +319,21 @@ static const struct of_device_id m25p_of_table[] = {
+ };
+ MODULE_DEVICE_TABLE(of, m25p_of_table);
+ 
++#ifdef CONFIG_ACPI
++static const struct acpi_device_id m25p_acpi_table[] = {
++	{ "JEDEC,SPI-NOR", 0 },
++	{ "ACPI0000", 0 },
```
Comment (lguohan) [inline patch/cisco-acpi-spi-nor.patch:29]: is this a standard name? will this have any conflict with existing acpi device?

Comment (lguohan) [reply patch/cisco-acpi-spi-nor.patch:29]: @msiddare to evaluate whether ACPI0000 is still needed or not.

**Comment by anshuv-mfst** [comment]
@paulmenzel , @lguohan - could you please review approve if comments are addressed?

File: patch/cisco-acpi-spi-nor.patch (hunk 5 above)
Comment (paulmenzel) [reply patch/cisco-acpi-spi-nor.patch:29]: This is still unanswered.

Comment (msiddare) [reply patch/cisco-acpi-spi-nor.patch:29]: Hi Paul,

"ACPI0000" is required for the detecting the MTD partition. 
I have experimented removing "ACPI000" to check if MTD partitions are recognized by driver during probe and they are not. 
This patch is only for 4.19.152 kernel, will port changes to 5.10 according to new file.

Thanks
Madhava

**Review by lguohan — APPROVED** [review]

**Comment by lguohan** [comment]
@msiddare, can you check the build failure?

**Comment by msiddare** [comment]
Hi Guohan,

There is an update to the patch which fixed this build failure.
Error is due to this: ACPI_PTR(m25p_acpi_table)

It needs an include file, which is resolved locally, I taught I pushed it already. seems like not. 
Will update new patch

**Pull request closed without merging**